File: solutions/python/binary-search-tree/1/binary_search_tree.py
class TreeNode:
    def __init__(self, data, left=None, right=None):
        self.data = data
        self.left = left
        self.right = right

    def __str__(self):
        return f'TreeNode(data={self.data}, left={self.left}, right={self.right})'


class BinarySearchTree:
    def __init__(self, tree_data):
        self.root = None

        for data in tree_data:
            self.insert(data, self.root)

    def data(self):
        return self.root

    def sorted_data(self):
        stack = []
        node = self.root
        data = []

        while node is not None or stack:
            while node is not None:
                stack.append(node)
                node = node.left

            node = stack.pop()
            data.append(node.data)

            node = node.right

        return data

    def insert(self, data, cur_node):
        if self.root is None:
            self.root = TreeNode(data)
            return

        if data <= cur_node.data:
            if cur_node.left is None:
                cur_node.left = TreeNode(data)
            else:
                self.insert(data, cur_node.left)
        else:
            if cur_node.right is None:
                cur_node.right = TreeNode(data)
            else:
                self.insert(data, cur_node.right)

File: solutions/c/binary-search-tree/1/binary_search_tree.c
#include "binary_search_tree.h"

#include <stdlib.h>

static void insert_data(int data, node_t **node) {
    if (*node == NULL) {
        *node = calloc(1, sizeof(node_t));
        (*node)->data = data;
    } else if (data <= (*node)->data) {
        insert_data(data, &((*node)->left));
    } else {
        insert_data(data, &((*node)->right));
    }
}

static void build_sorted_data(node_t *node, int *sorted_list, int *index) {
    if (node) {
        build_sorted_data(node->left, sorted_list, index);
        sorted_list[(*index)++] = node->data;
        build_sorted_data(node->right, sorted_list, index);
    }
}

node_t *build_tree(int *tree_data, size_t tree_data_len) {
    node_t *root = NULL;

    for (size_t i = 0; i < tree_data_len; i++) {
        insert_data(tree_data[i], &root);
    }

    return root;
}

void free_tree(node_t *tree) {
    if (!tree) return;

    free_tree(tree->left);
    free_tree(tree->right);
    free(tree);
}

void tree_size(node_t *node, int *count) {
    if (node) {
        tree_size(node->left, count);
        (*count)++;
        tree_size(node->right, count);
    }
}

int *sorted_data(node_t *tree) {
    int count = 0;
    tree_size(tree, &count);

    int *sorted_list = calloc(count, sizeof(int));
    int index = 0;
    build_sorted_data(tree, sorted_list, &index);

    return sorted_list;
}